tools/hdr/local_tone_map.cc

```diff
@@ -0,0 +1,378 @@
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT input_rows[3] = {
+        image.PlaneRow(0, y), image.PlaneRow(1, y), image.PlaneRow(2, y)};
+    float* const JXL_RESTRICT row = grayscale.Row(y);
+
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      row[x] = LinearToSrgb8Direct(
+          kLuminances[0] * Srgb8ToLinearDirect(input_rows[0][x]) +
+          kLuminances[1] * Srgb8ToLinearDirect(input_rows[1][x]) +
+          kLuminances[2] * Srgb8ToLinearDirect(input_rows[2][x]));
+    }
+  }
+
+  ImageF laplacian(image.xsize(), image.ysize());
+  Symmetric3(grayscale, Rect(grayscale), kLaplacianWeights, pool, &laplacian);
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT row = laplacian.ConstRow(y);
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      consumer(x, y, std::abs(row[x]));
+    }
+  }
+}
+
+template <typename Consumer>
+void Saturation(const jxl::Image3F& image, const Consumer& consumer) {
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT rows[3] = {
+        image.PlaneRow(0, y), image.PlaneRow(1, y), image.PlaneRow(2, y)};
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      const float mean = (1.f / 3) * (rows[0][x] + rows[1][x] + rows[2][x]);
+      const float deviations[3] = {rows[0][x] - mean, rows[1][x] - mean,
+                                   rows[2][x] - mean};
+      consumer(x, y,
+               std::sqrt((1.f / 3) * (deviations[0] * deviations[0] +
+                                      deviations[1] * deviations[1] +
+                                      deviations[2] * deviations[2])));
+    }
+  }
+}
+
+template <typename Consumer>
+void MidToneness(const jxl::Image3F& image, const float sigma,
+                 const Consumer& consumer) {
+  const float inv_sigma_squared = 1.f / (sigma * sigma);
+  const auto Gaussian = [inv_sigma_squared](const float x) {
+    return std::exp(-.5f * (x - .5f) * (x - .5f) * inv_sigma_squared);
+  };
+  for (size_t y = 0; y < image.ysize(); ++y) {
+    const float* const JXL_RESTRICT rows[3] = {
+        image.PlaneRow(0, y), image.PlaneRow(1, y), image.PlaneRow(2, y)};
+    for (size_t x = 0; x < image.xsize(); ++x) {
+      consumer(
+          x, y,
+          Gaussian(rows[0][x]) * Gaussian(rows[1][x]) * Gaussian(rows[2][x]));
+    }
+  }
+}
+
+ImageF ComputeWeights(const jxl::Image3F& image, const float contrast_weight,
+                      const float saturation_weight,
+                      const float midtoneness_weight,
+                      const float midtoneness_sigma, ThreadPool* const pool) {
+  ImageF weights(image.xsize(), image.ysize());
+  FillImage(1.f, &weights);
+
+  if (contrast_weight > 0) {
+    Contrast(
+        image,
+        [&weights, contrast_weight](const size_t x, const size_t y,
+                                    const float weight) {
+          weights.Row(y)[x] = std::pow(weight, contrast_weight);
```

Comment:
A small optimization (also slightly numerically more sound) would be to fill with 0, add log(weight) * contrast_weight, and then exp in the end.

Reply:
Oops, I resolved and did the change locally but did not push. One moment…